zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -160,35 +258,28 @@ pub struct LtGadget<F, const NUM_BYTES: usize> {
 }
 
 impl<F: FieldExt, const NUM_BYTES: usize> LtGadget<F, NUM_BYTES> {
-    pub const NUM_CELLS: usize = 1 + NUM_BYTES; // lt + diff
-    pub const NUM_WORDS: usize = 0;
-
-    pub(crate) fn construct(alloc: &mut CaseAllocation<F>) -> Self {
-        assert!(NUM_BYTES <= MAX_BYTES_FIELD, "unsupported number of bytes");
-        Self {
-            lt: alloc.cells.pop().unwrap(),
-            diff: array_init(|_| alloc.cells.pop().unwrap()),
-            range: get_range(NUM_BYTES * 8),
-        }
-    }
-
-    pub(crate) fn constraints(
-        &self,
+    pub(crate) fn construct(
         cb: &mut ConstraintBuilder<F>,
         lhs: Expression<F>,
         rhs: Expression<F>,
-    ) -> Expression<F> {
-        let diff = from_bytes::expr(self.diff.to_vec());
+    ) -> Self {
+        assert!(NUM_BYTES <= MAX_BYTES_FIELD, "unsupported number of bytes");
+
+        let lt = cb.query_bool();
+        let diff = cb.query_bytes();
+        let range = pow_of_two(NUM_BYTES * 8);
+
         // The equation we require to hold: `lhs - rhs == diff - (lt * range)`.
```

Comment:
why not `lhs - rhs == diff + (lt * range) ?

Reply:
When `lt`, the `lhs - rhs` would be negative value, which never fits in the `NUM_BYTES`. So adding it with the `range` we are targeting should makes `lhs - rhs + range` fit again. The equation here just moves the `+ range` to the right hand side, so it becomes `lhs - rhs == diff - (lt ⋅ range)`

Reply:
oh, it is the moving , right !  thanks